src/cache.rs

```diff
@@ -0,0 +1,305 @@
+      })
+      .collect::<Result<Vec<CircuitShape<G1>>, _>>();
+
+    let circuit_digests = CircuitDigests::<G1>::new(digests);
+    let maybe_aux_params = self.get::<G1::Scalar, AuxParams<G1, G2>>(circuit_digests.digest());
+
+    let pp = match (maybe_circuit_shapes, maybe_aux_params) {
+      (Ok(circuit_shapes), Ok(aux_params)) => {
+        tracing::info!("cache hit");
+
+        let size_primary = circuit_shapes
+          .iter()
+          .map(|circuit| commitment_key_size(&circuit.r1cs_shape, None))
+          .max()
+          .unwrap();
+        let ck_primary = self.commitment_key::<G1>(size_primary)?;
+
+        PublicParams::<G1, G2, C1, C2>::from_parts_unchecked(circuit_shapes, aux_params, ck_primary)
+      }
+      _ => {
+        tracing::info!("generating public params");
+        let pp = PublicParams::new(non_uniform_circuit);
+
+        let (circuit_shapes, aux_params, ck_primary) = pp.into_parts();
+
+        self.set::<G1::Scalar, AuxParams<G1, G2>>(circuit_digests.digest(), &aux_params)?;
+        for (digest, circuit_shape) in circuit_digests.deref().iter().zip(circuit_shapes.iter()) {
+          self.set::<G1::Scalar, CircuitShape<G1>>(*digest, circuit_shape)?;
+        }
+        self.set_commitment_key::<G1>(&ck_primary)?;
+
+        PublicParams::<G1, G2, C1, C2>::from_parts_unchecked(circuit_shapes, aux_params, ck_primary)
+      }
+    };
+
+    Ok(pp)
+  }
+
+  fn get<F: PrimeField, T: Abomonation + Clone>(&self, digest: F) -> io::Result<T> {
+    let (file, exists) = self.digest_file::<F>(digest)?;
+
+    if exists {
+      unsafe {
+        let mut mmap = MmapMut::map_mut(&file)?;
+        let (t, remaining) = abomonation::decode::<T>(&mut mmap).unwrap();
+        assert!(remaining.is_empty());
+        Ok(t.clone())
+      }
+    } else {
+      Err(Error::new(
+        ErrorKind::Other,
+        "failed to retrieve from cache",
+      ))
+    }
+  }
+
+  fn set<F: PrimeField, T: Abomonation>(&self, digest: F, t: &T) -> io::Result<()> {
+    let (file, exists) = self.digest_file::<F>(digest)?;
+
+    if exists {
+      tracing::info!("object already in cache; nothing will be written");
+    } else {
+      let mut writer = BufWriter::new(file);
+      unsafe { encode(t, &mut writer)? };
+    }
+
+    Ok(())
+  }
+
+  /// Deletes all the commitment keys in the cache
+  ///
+  /// **Warning**: This will indiscriminately wipe out everything in `<cache>/commitment_keys`,
+  /// so don't call this unless you're **sure** that there's nothing important in there.
+  pub fn clear_commitment_keys(&self) -> io::Result<()> {
+    let path = self.get_inner().join("commitment_keys");
+    remove_dir_all(&path)?;
+    create_dir_all(path)
+  }
+}
+
+#[cfg(test)]
+mod tests {
+  use std::path::PathBuf;
+
+  use pasta_curves::pallas;
+
+  use crate::{
+    provider::{bn256_grumpkin::bn256, secp_secq::secq256k1},
+    traits::{commitment::CommitmentEngineTrait, Group},
+  };
+
```

Comment:
I think the concurrency test is three threads all trying to extend the same CK, here.